tools/ts-generator/types/types_gen_js.py

```diff
@@ -246,24 +246,36 @@
 
 {%- macro read_array(type, buffer, offset, func, size) -%}
     {# Remove ">" and "Array<" from type, the result is the array type #}
-    {%-set subtype = type | replace(">","")|replace("Array<", "") -%}
+    {%-set subtype = type | get_subtype -%}
+    {%- if func == False -%}
     (() => {
         const [array, newOffset] = BF.readArray({{size}})({{buffer}}, {{offset}}, 
         {{- deserialize_by_type({"type": subtype}, "auxBuffer", "auxOffset", True) -}})
         offset = newOffset
         return array;
     })()
+    {%- else -%}
+        ({{buffer}}, {{offset}}) => 
+        BF.readArray({{size}})({{buffer}}, {{offset}}, 
+            {{- deserialize_by_type({"type": subtype}, "auxBuffer", "auxOffset", True) -}})
+    {%- endif -%}
 {%- endmacro %}
 
 {%- macro read_optional(type, buffer, offset, func) -%}
     {# Remove ">" and "Optional<" from type, the result is the optional type #}
-    {%-set subtype = type | replace(">","") | replace("Optional<", "") -%}
+    {%-set subtype = type | get_subtype -%}
+    {%- if func == False -%}
     (() => {
         const [optional, newOffset] = BF.readOptional({{buffer}}, {{offset}},
         {{- deserialize_by_type({"type": subtype}, "auxBuffer", "auxOffset", True) -}})
         offset = newOffset
         return optional;
     })()
+    {%- else -%}
+    ({{buffer}}, {{offset}}) => 
+        BF.readOptional({{buffer}}, {{offset}},
+        {{- deserialize_by_type({"type": subtype}, "auxBuffer", "auxOffset", True) -}})
```

Comment:
Will this work for deeply nested types? Like `Optional<Array<Optional<Optional<T>>>>` for example?

Reply:
I guess subtype is correctly applied at each level

Reply:
https://github.com/vectorizedio/redpanda/pull/696/files#diff-024b655d0230fc6f6090500a4141f50d289678d6e25356105522fc731a65ebb0R78

done I added this type to demo.ts